### PR TITLE
[rbd] support G/T units in rbd create/resize

### PR DESCRIPTION
**Pull request by vumrao**
Hi Josh,

Please review this patch. I have done basic unit testing and it is working fine. 
Still need to modify doc and man pages and release notes.

Thanks and Regards,
Vikhyat

### CODE REVIEW

**Comment by yuyuyu101** [comment]
plz add Signed-off-by: via "git -s ..."

File: src/rbd.cc [hunk 1]
```diff
@@ -98,14 +98,16 @@ void usage()
 "  info <image-name>                           show information about image size,\n"
 "                                              striping, etc.\n"
 "  create [--order <bits>] [--image-features <features>] [--image-shared]\n"
-"         --size <MB> <name>                   create an empty image\n"
+"         --size <value> --sizeunit <M/G/T>\n"
+"         <image-name>                         create an empty image\n"
```
Comment (yuyuyu101) [inline src/rbd.cc]: we may more like to use this way :

rbd create image -p pool --size 100G

**Comment by vumrao** [comment]
Thanks Haomai for your inputs, I will make the changes and will upload the new patchset.
Regards,
Vikhyat

**Comment by vumrao** [comment]
Sorry I think I am missing something here in github , how to send multiple patchset here and because of that mistakes three commits went blank.

Correct commit is : 1d245428f01bb9c662f4df89dab567c5e3d90ceb , please review this commit.

Thanks and Regards,
Vikhyat

**Comment by yuyuyu101** [comment]
Ah, I'm not sure whether you done for this branch. Maybe you need to git push -f?

**Comment by theanalyst** [comment]
shoudn't this default to MB as well ie. even if M/MB isn't specified to be compatible with current defaults?

**Comment by vumrao** [comment]
Hi Abhishek,

Thanks for the review, else part handles that case. 

Regards,
Vikhyat

**Comment by theanalyst** [comment]
Vikhyat, thanks for the clarification (I missed the else clause), to clean up the empty commits you could rebase against master & do a force push of this branch

**Comment by vumrao** [comment]
Thanks Abhishek for your inputs, yes I did it and now final commit is bb4924b.

**Comment by liewegas** [comment]
We have lots of places in the code where we specify sizes, and with a range of different units (bytes, MB, mostly).  I would prefer we create a generic converter/parser function, probably in common/strtol .. maybe strtoll_units?  And we can make it take an argument with the default units if they are not explicitly specified.  Then we have one function, with unit tests, to maintain.

Eventually we can fix up the config parsing to make use of this (by including a units field with each option).

**Comment by vumrao** [comment]
1. Thank you Sage for your inputs, sure I will work on it for having a common converter/parser function like strtoll_units.

2. Coming to your this suggestion - 

> Eventually we can fix up the config parsing to make use of this (by including a units field with each option).
Are you suggesting something like which I have done in my first commit of this pull request : 

https://github.com/vumrao/ceph/commit/718c41c6ec77e703557de6d3d59d52a2a5cb4fda 
I have added --sizeunit <M/G/T> in this .

Please let me know your inputs.

Thanks and Regards,
Vikhyat

**Comment by liewegas** [comment]
In the rbd CLI case we should just use strict_strtoll_units (or whatever).  For config parsing I'm referring to the options in common/config_opts.h (and parsing in common/config.cc).  That's a bigger project, though, for later!

**Comment by vumrao** [comment]
Thanks Sage.
For rbd size and resize G/T support , I just checked and found in common/strtol.cc we have already one function which takes care of conversion and parsing strict_sistrtoll() and I am making use of it.
Currently I have added one condition to keep rbd size/resize behavior so all applications which use this CLI will not break and once they will adopt B/K/M/G/T/P/E with size value we can remove this default if condition.

**Comment by jdurgin** [comment]
Looks good! Could you also update the man page (doc/man/8/rbd.rst) with the accepted size units, and one more test (src/test/cli/rbd/help.t) with the new help output)?

The commits should be squashed so there's only one - you can do this by running git rebase -i 718c41c6ec77e703557de6d3d59d52a2a5cb4fda^

and changing pick to squash for the 2nd commit.
Thanks!

**Comment by vumrao** [comment]
Hi Josh,

Thanks for your inputs.
I have made the changes in my first commit and pushed the changes as I was getting lots of issues while squashing both the commits as I have created my second commit on top of first commit.

Please let me know your inputs.

Thanks Kefu for your help on working with me resolving git issue finally I have made the changes on my first commit and push with -f.

Regards,
Vikhyat

**Comment by dachary** [comment]
@dillaman another run that may interest you

**Comment by vumrao** [comment]
Thanks Every one !

**Comment by dillaman** [comment]
@dachary thanks for the ping.  it looks like a repeat of #4912

**Pull request merged**